### DISCUSSION TITLE
How to implement like show tooltips when mouse hover some specific widget?

### DISCUSSION THREAD

**Discussion by mhfan** [slint-ui/slint#1617] [Q&A]
I guess a popupwindow might be a proper way, but how to bind the "show()" show to "has-hover" property of a touch area?

**Comment by ogoffart** [comment]
Tooltip is one of the things that is not yet implemented.

Ideally, we'd have something like a `tooltip`  property that could be applied to any element and would magcally work
```
    Button {
        tooltip: "This button does usefull things";
        //...
    }
```

But since we don't have that, the PopupWindow is a second good choice.  Unfortunately, the PopupWindow is a bit basic and it steals the mouse input in the whole window, which is also not what we want.

So we can come up with this hack, right now.
[link](https://slint-ui.com/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Aglobal+CurrentToolTip+%3A%3D+%7B%0A++++property+%3Cstring%3E+text%3B%0A++++property+%3Cbool%3E+visible%3B%0A++++property+%3CPoint%3E+pos%3B%0A%7D%0A%0AToolTipArea+%3A%3D+TouchArea+%7B%0A++++property+%3Cstring%3E+text%3B%0A++++property+%3CPoint%3E+offset%3B%0A++++%0A++++property+%3Cbool%3E+_active%3A+false%3B%0A%0A%0A++++callback+set_active%3B%0A++++set_active+%3D%3E+%7B%0A++++++++CurrentToolTip.pos+%3D+%7B+%0A++++++++++++x%3A+mouse-x+%2B+root.x+%2B+offset.x%2C+%0A++++++++++++y%3A+mouse-y+%2B+root.y+%2B+offset.y%2C%0A++++++++%7D%3B%0A++++++++CurrentToolTip.text+%3D+root.text%3B%0A++++++++CurrentToolTip.visible+%3D+true%3B%0A++++++++_active+%3D+true%3B%0A++++%7D%0A%0A++++moved+%3D%3E+%7B+%0A+++++++%2F%2F+set_active%28%29%3B%0A++++%7D%0A++++pointer-event%28event%29+%3D%3E+%7B+%0A++++++++if+%28_active+%26%26+event.kind+%3D%3D+PointerEventKind.cancel%29+%7B%0A++++++++++++CurrentToolTip.visible+%3D+false%3B%0A++++++++++++_active+%3D+false%3B%0A++++++++%7D%0A++++%7D%0A%0A++++%2F%2F+Hack%3A+side+effect+in+property+access%0A++++enabled%3A+%7B%0A++++++++if+%28has-hover%29+%7B%0A++++++++++++set_active%28%29%3B%0A++++++++%7D+else+if+%28_active%29+%7B%0A++++++++++++CurrentToolTip.visible+%3D+false%3B%0A++++++++++++_active+%3D+false%3B%0A++++++++%7D%0A++++++++return+true%3B%0A++++%7D%0A%0A%7D%0A%0Aexport+Demo+%3A%3D+Window+%7B%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++++++ToolTipArea+%7B+%0A++++++++++++++++text%3A+%22Some+Text+with+Hello+World%22%3B+%0A++++++++++++++++offset%3A+%7B+x%3A+parent.x%2C+y%3A+parent.y+%7D%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++Image+%7B%0A++++++++++++source%3A+%40image-url%28%22https%3A%2F%2Fslint-ui.com%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++height%3A+100px%3B%0A++++++++++++ToolTipArea+%7B+%0A++++++++++++++++text%3A+%22An+Image+of+the+Slint+Logo%22%3B+%0A++++++++++++++++offset%3A+%7B+x%3A+parent.x%2C+y%3A+parent.y+%7D%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++hl+%3A%3D+HorizontalLayout+%7B+%0A++++++++++++alignment%3A+center%3B+%0A++++++++++++Button+%7B+%0A++++++++++++++++text%3A+%22OK%21%22%3B+%0A++++++++++++++++ToolTipArea+%7B+%0A++++++++++++++++++++text%3A+%22Button+to+accept%22%3B+%0A++++++++++++++++++++offset%3A+%7B+x%3A+parent.x+%2B+hl.x%2C+y%3A+parent.y+%2B+hl.y+%7D%3B%0A++++++++++++++++%7D%0A++++++++++++%7D+%0A++++++++%7D%0A++++%7D%0A%0A++++Rectangle+%7B+%0A++++++++visible%3ACurrentToolTip.visible%3B%0A++++++++states+%5B+%0A++++++++++++visible+when+visible%3A+%7B+opacity%3A+0.8%3B+%7D%0A++++++++%5D%0A++++++++transitions+%5B+%0A++++++++++++in+visible+%3A+%7B%0A++++++++++++++++animate+opacity+%7B+duration%3A+175ms%3B+delay%3A+700ms%3B+%7D%0A++++++++++++%7D%0A++++++++%5D%0A++++++++x%3A+CurrentToolTip.pos.x%3B%0A++++++++y%3A+CurrentToolTip.pos.y%3B%0A++++++++background%3A+yellow%3B%0A++++++++opacity%3A+0%3B%0A++++++++width%3A+tt_l.preferred-width%3B%0A++++++++height%3A+tt_l.preferred-height%3B%0A++++++++tt_l+%3A%3D+HorizontalLayout+%7B+%0A++++++++++++padding%3A+3px%3B%0A++++++++++++Text+%7B+text%3A+CurrentToolTip.text%3B+%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A)

It's a bit hackish but can be polished somehow.


Long term we would like to have a proper solution within Slint of course.

**Reply by ogoffart** [reply]
Actually, i came up with a less hackish solution: [link](https://slint-ui.com/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0AToolTipArea+%3A%3D+TouchArea+%7B%0A++++property+%3Cstring%3E+text%3B%0A++++%0A++++Rectangle+%7B+%0A++++++++visible%3A+parent.has-hover%3B%0A++++++++states+%5B+%0A++++++++++++visible+when+visible%3A+%7B+opacity%3A+0.8%3B+%7D%0A++++++++%5D%0A++++++++transitions+%5B+%0A++++++++++++in+visible+%3A+%7B%0A++++++++++++++++animate+opacity+%7B+duration%3A+175ms%3B+delay%3A+700ms%3B+%7D%0A++++++++++++%7D%0A++++++++%5D%0A++++++++x%3A+parent.mouse-x%3B%0A++++++++y%3A+parent.mouse-y%3B%0A++++++++background%3A+yellow%3B%0A++++++++opacity%3A+0%3B%0A++++++++width%3A+tt_l.preferred-width%3B%0A++++++++height%3A+tt_l.preferred-height%3B%0A++++++++tt_l+%3A%3D+HorizontalLayout+%7B+%0A++++++++++++padding%3A+3px%3B%0A++++++++++++Text+%7B+text+%3C%3D%3E+root.text%3B+%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A%0Aexport+Demo+%3A%3D+Window+%7B%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++++++ToolTipArea+%7B+%0A++++++++++++++++text%3A+%22Some+Text+with+Hello+World%22%3B+%0A++++++++++++%7D%0A++++++++%7D%0A++++++++Image+%7B%0A++++++++++++source%3A+%40image-url%28%22https%3A%2F%2Fslint-ui.com%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++height%3A+100px%3B%0A++++++++++++ToolTipArea+%7B+%0A++++++++++++++++text%3A+%22An+Image+of+the+Slint+Logo%22%3B+%0A++++++++++++%7D%0A++++++++%7D%0A++++++++hl+%3A%3D+HorizontalLayout+%7B+%0A++++++++++++alignment%3A+center%3B+%0A++++++++++++Button+%7B+%0A++++++++++++++++text%3A+%22OK%21%22%3B+%0A++++++++++++++++ToolTipArea+%7B+%0A++++++++++++++++++++text%3A+%22Button+to+accept%22%3B+%0A++++++++++++++++%7D%0A++++++++++++%7D+%0A++++++++%7D%0A++++%7D%0A++%0A%7D%0A)

But it may actually not work if the tooltip gets under another item

**Reply by mhfan** [reply]
It  works for tooltip; However it seems this solution made Button no responsive to click/hover actions, is it solvable?

Besides, when there are several buttons, the tooltip area of some button will be shown wired under other buttons...

**Comment by ogoffart** [comment]
Still no supported builtin tooltip in Slint.
But here is a refreshed code updated to Slint 1.0

[SlintPad link](https://slintpad.com/?snippet=component+ToolTipArea+%7B%0A++++preferred-height%3A+100%25%3B%0A++++preferred-width%3A+100%25%3B%0A++++%0A++++in+property+%3Cstring%3E+text%3B%0A%0A++++ta+%3A%3D+TouchArea+%7B%7D%0A++++%0A++++Rectangle+%7B+%0A++++++++states+%5B+%0A++++++++++++visible+when+ta.has-hover%3A+%7B+%0A++++++++++++++++opacity%3A+0.8%3B+%0A++++++++++++++++in+%7B+%0A+++++++++++++++++++animate+opacity+%7B+duration%3A+175ms%3B+delay%3A+700ms%3B+%7D%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%5D%0A++++++++x%3A+ta.mouse-x+%2B+1rem%3B%0A++++++++y%3A+ta.mouse-y+%2B+1rem%3B%0A++++++++background%3A+lightyellow%3B%0A++++++++opacity%3A+0%3B%0A++++++++width%3A+tt_l.preferred-width%3B%0A++++++++height%3A+tt_l.preferred-height%3B%0A++++++++tt_l+%3A%3D+HorizontalLayout+%7B+%0A++++++++++++padding%3A+3px%3B%0A++++++++++++Text+%7B+text+%3C%3D%3E+root.text%3B+%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A%0Aexport+component+Demo+inherits+Window+%7B%0A++++width%3A+200px%3B%0A++++height%3A+200px%3B%0A++++ToolTipArea+%7B+text%3A+%22This+is+a+tooltip%22%3B+%7D%0A%7D%0A)

```slint
component ToolTipArea {
    preferred-height: 100%;
    preferred-width: 100%;
    
    in property <string> text;

    ta := TouchArea {}
    
    Rectangle { 
        states [ 
            visible when ta.has-hover: { 
                opacity: 0.8; 
                in { 
                   animate opacity { duration: 175ms; delay: 700ms; }
                }
            }
        ]
        x: ta.mouse-x + 1rem;
        y: ta.mouse-y + 1rem;
        background: lightyellow;
        opacity: 0;
        width: tt_l.preferred-width;
        height: tt_l.preferred-height;
        tt_l := HorizontalLayout { 
            padding: 3px;
            Text { text <=> root.text; }
        }
    }
}
```

**Comment by ogoffart** [comment]
There is now an issue with another example: https://github.com/slint-ui/slint/issues/6446#issuecomment-2394554920
